### PR TITLE
fix(marker): keep the child parse bound inside the bounds wrapping it

### LINKED ISSUE #894: parse_timeout and marker_timeout share a default, so the #659 pool recycle never runs
## Problem

`parse_timeout` and `marker_timeout` both default to **3600**, and the parse-stage
bound starts *earlier* than Marker's own. The outer therefore always expires first,
so the only code path that recycles a wedged Marker pool (#659) is unreachable.

Three nested timeouts, all `3600`:

| # | Where | Value | Clock starts | Recycles the pool? |
|---|---|---|---|---|
| 1 | `services/workers/stages/parse.py:43` | `loader.parse_timeout` | parse-stage entry | no |
| 2 | `services/workers/parsers/marker_workers.py:326` (`_run_chunk`) | `loader.marker_timeout` | after slot acquisition | no |
| 3 | `services/workers/parsers/marker_workers.py:210` (`process_pdf`) | `loader.marker_timeout` | at child `submit()` | **yes** |

Only #3 recycles:

```python
except FuturesTimeoutError:
    # ... recycle the pool to reclaim the wedged worker's slot (#659)
    self.setup_mp(old_executor=current_executor)
```

Because #1 starts first and all three share a duration, #1 wins for any non-zero
slot wait, cancelling the chain before #3's handler runs.

This contradicts the documented intent in `conf/config.yaml`:

> Outer wall-clock bound (seconds) for a single file's parse stage. **marker and
> docling self-limit at marker_timeout/docling_timeout**; pymupdf has no internal
> timeout […]

`parse_timeout` is meant to be a backstop for PyMuPDF, with Marker self-limiting
first. The shared default inverts that.

## Impact

A wedged Marker worker is never reclaimed. It keeps its pool slot and the GPU, and
every subsequent document fails at the outer timeout. Only a manual restart recovers.

Observed in a production deployment on `v2.1.1`: a batch of PDFs (165 KB–2.2 MB)
submitted within ~400 ms of each other **all failed at exactly 3600 s**
(`duration_ms` 3600031–3600218). After a plain stack restart with unchanged
configuration, the same 40 documents indexed in **4 minutes** (~6 s/document) —
confirming the documents were fine and the pool was stalled.

The traceback shows the inner call was **cancelled from outside**, not timed out —
`asyncio.exceptions.CancelledError` propagating into
`TimeoutError: parse timed out after 3600s`, with no
`"Resetting ProcessPoolExecutor"` log line. The recycle never ran.

## Fix

Enforce the ordering invariant rather than documenting it:

```
parse_timeout > marker_timeout + max_expected_slot_wait
```

Suggested:
- Validate at startup and refuse (or warn loudly) when `parse_timeout <= marker_timeout`,
  mirroring the existing `ALLOW_NO_AUTH` gating pattern.
- Consider deriving `parse_timeout` from `marker_timeout` (e.g. ×2) instead of
  defaulting both to 3600.
- Same relationship applies to `docling_timeout`.

## Workaround

Config-only, no code change:

```
MARKER_TIMEOUT=900
PARSE_TIMEOUT=3600
```

The inner bound then fires first and the pool recycles.

## Related

- #659 — the recycle this defect makes unreachable (closed, code is correct)
- #723 — the caller-cancelled slot-release hole; same cancellation mechanism, different defect

## Notes

`tests/unit/services/workers/parsers/test_marker_workers.py` exercises `process_pdf`
directly, so #3 fires in isolation and passes. The defect lives in the composition
of the three layers; a regression test needs to drive stage → pool → actor.

Verified on `origin/develop` @ `5687cd74`.

### PR DESCRIPTION
**Pull request by EnjoyBacon7**
Closes #894

Four timeouts wrap a single Marker parse: the parse stage, the loader, the pool,
and the child process. All of them default to 3600s.

Only the child one recycles a wedged pool. It also starts its clock last, so one
of the outer three always ran out first and cancelled everything before the
recycle could happen. The stuck worker then kept its slot and the GPU until
someone restarted the stack.

The child bound is now a fraction of `marker_timeout` (0.9 by default, so 3240s),
which makes it expire first no matter how the outer ones are configured.

## Summary by CodeRabbit

- **New Features**
  - Added configurable time allocation for Marker child-process operations.
  - Added support for the `MARKER_CHILD_TIMEOUT_RATIO` environment variable, defaulting to 90%.
  - Child-process timeouts now remain within the overall Marker and parsing time limits.

- **Bug Fixes**
  - Improved timeout handling so child operations can time out cleanly before surrounding processing is terminated.
  - Added validation to reject invalid Marker timeout and child-timeout ratio values.

### CODE REVIEW

File: openrag/core/config/indexation.py [hunk 1]
```diff
@@ -178,6 +178,21 @@ class LoaderConfig(ConfigMixin):
...
+    # pool calls, both ``marker_timeout``, and the parse stage's
+    # ``parse_timeout``) all start their clock *earlier*. Sharing one duration
+    # therefore let an outer bound cancel the chain at or before the moment the
+    # recycle would fire, so a wedged worker kept its slot and the GPU until the
+    # stack was restarted (#894).
+    #
+    # Scaling the child bound below ``marker_timeout`` restores the ordering by
+    # construction, independent of how the outer bounds are configured. The
+    # remaining fraction is the headroom that ordering needs.
+    marker_child_timeout_ratio: float = Field(default=0.9, gt=0, lt=1)
```
[the remark this answers is not in the document]
Comment (EnjoyBacon7) [reply openrag/core/config/indexation.py:185]: Fixed in 667a06f1. The child bound is now `min(marker_timeout, parse_timeout) * ratio`, so the shortest wrapping bound sets the ceiling. Defaults are unchanged at 3240s. Added a regression test for `parse_timeout=60` with `marker_timeout=3600`, which now yields 54s; it fails against the previous property.

One correction on the mechanism: the recycle was not skipped, it was delayed. `run_in_executor` returns a future that cancels the awaiting coroutine but not the worker thread, so `run_with_timeout` kept running and reached `setup_mp` at 0.9 * marker_timeout. The effect you describe still holds, since the slot stays wedged for that whole period and every queued document fails meanwhile.

**Comment by hedhoud** [comment]
The original child-timeout issue is resolved. `marker_child_timeout` now stays below both `marker_timeout` and `parse_timeout`, and the shortened `parse_timeout` case is covered by a regression test.

One issue remains before approval:

- In `openrag/core/config/indexation.py:180`, `marker_timeout` accepts zero or negative values because it has no positive-value validation.
- In `openrag/core/config/indexation.py:215`, this can produce a zero or negative `marker_child_timeout`, violating the timeout-ordering guarantee.

Please require `marker_timeout` to be greater than zero and add focused tests for zero and negative values in `tests/unit/core/config/test_indexation.py`. No critical issues were found; approval can follow once this fix is implemented and CI passes.

**Comment by EnjoyBacon7** [comment]
Fixed in 58db8c6e. `marker_timeout` now requires `gt=0`, so a zero or negative value is rejected at config load instead of producing a non-positive child bound.
Regression test covers both 0 and -1.
Worth noting the field was already unvalidated on develop, where the same value breaks the four `call_ray_actor_with_timeout` sites in `marker_workers.py`, so this is a hardening rather than a regression from this PR.

**Review by hedhoud — APPROVED** [review]
Reviewed the changes. No actionable issues found, and the targeted and full unit suites pass.
